rhizome/bin/prep_host.rb

```diff
@@ -36,3 +46,11 @@
 # For qemu-image convert and mcopy for cloud-init with the nocloud
 # driver.
 r "apt-get -y install qemu-utils mtools"
+
+# We currently use 1GB of hugepages per VM, so by requesting 16K 2MB pages as
+# below, we will have enough hugepages for 32 VMs in a host.
+#
+# TODO: It is possible that OS allocates less hugepages than requested, so
+# check how much hugepages were actually allocated, and use it in capacity
+# calculations.
+r "echo 16384 | sudo tee /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages"
```

Comment:
A gigabyte of overhead in queues per VM seems awfully steep unless we're talking absolutely ginormous VMs pushing incredible storage traffic. How well does it work if you allocate, like, a few...maybe even one or two...2M hugepages per VM, or per VM-core? 

Another question I have is if the queue space is shared between multiple storage devices, e.g. if a VM has 20 virtual PCI storage devices/virtio-blk instantiations, does it need at least 20 hugepages? This is kind of a detailed question about how virtio-blk and vhost works, that I don't know the answer to right off. Or is virtio-blk/vhost smart enough to take, say, two hugepages (4MB) and divide it up best it can between all the virtual PCI devices?

This is of a little importance, because it would be simpler to decide bigger VMs probably want to do more I/O and scale queue size relative to cores, rather than being caught with something more difficult depending on how many I/O devices are virtualized within.

Reply:
I'm still looking into details of your questions in last 3 paragraphs.

But this is not the memory just for queues, but for everything SPDK related. SPDK uses DPDK for memory allocation (see implementation of spdk_malloc & spdk_zmalloc [1]), and DPDK's memory allocator uses a pool of hugepages [2]. 

I experimented with lowering the memory size, and SPDK's vhost target could boot with 400MB of memory, so I lowered the 1GB memory / VM to 512M memory / VM for now. I also posted a question about this in the SPDK slack group.

However we could use less memory per if we share the same vhost process between them. That might have some security risks, but I need to look into this more.

[1] https://github.com/spdk/spdk/blob/8301b0414a331e22cd7fbae683669a6a0394a5b9/lib/env_dpdk/env.c#L37
[2] https://www.dpdk.org/blog/2019/08/21/memory-in-dpdk-part-1-general-concepts/

Reply:
We talked a bit about this, summarizing the most important thoughts:

We're more or less obliged to share SPDK processes, because we want to multiplex a number of VMs onto storage hardware, and only one SPDK process can manipulate, say, a NVMe card at a time (since it bypasses the kernel). So, it's dangerous.

The things that are less clear is how much VM memory is exposed to SPDK. It looks like the hypervisor is actually responsible for telling spdk via socket what chunks of memory have been carved out for it, so as it turns out, those queues get carved out of the memory associated with the guest there. I asked the cloud-hypervisor slack about how much VM data is exposed. https://cloud-hypervisor.slack.com/archives/CJJ878PEY/p1686081030273339. Looking at the source, I think we could improve this if the answer is lacking, at some point.

The last question is: if the space for queues is carved out of each VM, what is SPDK doing with all that memory in its pool?